src/fa.c

```diff
@@ -4513,6 +4513,88 @@ void fa_dot(FILE *out, struct fa *fa) {
     fprintf(out, "}\n");
 }
 
+void fa_json(FILE *out, struct fa *fa) {
+    int *list_hashes;
+    int list_size = 100;
+    int num_states = 0;
+    int it;
+    char first = true;
+
+    fprintf(out,"{\n\t\"final\": [");
+
+    list_hashes = (int *)malloc(sizeof(int) * list_size);
+
+    list_for_each(s, fa->initial) {
+        if (num_states == list_size - 1){
+            list_size += list_size;
+            list_hashes = realloc(list_hashes, list_size);
+        }
+        // Store hash value
+        list_hashes[num_states] = s->hash;
+        // We use the hashes to map states to Z_{num_states}
+        s->hash = num_states++;
+        if (s->accept) {
+            if (!first) fprintf(out, ", %ld", s->hash);
+            else fprintf(out,"%ld", s->hash);
+            first = false;
+        }
+    }
+
+    fprintf(out, "],\n\t\"deterministic\": %d,\n\t\"transitions\": [\n", fa->deterministic ? 1 : 0);
+
+    first = true;
+    list_for_each(s, fa->initial) {
+        for_each_trans(t, s) {
+            if (!first) fprintf(out, ",\n");
+            first = false; 
+            fprintf(out, "\t\t{ \"from\": %ld, \"to\": %ld, \"on\": \"",s->hash, t->to->hash);
+            print_char(out, t->min);
+            if (t->min != t->max) {
+                fputc('-', out);
+                print_char(out, t->max);
+            }
+            fprintf(out, "\" }");
+        }
+    }
+
+    fprintf(out,"\n\t]\n}");
+
```

Comment:
This is probably where the `error` label should go (so that we restore hashes on error)